### ISSUE TITLE
IMPORTANT No way to use Input.Group and Form.Item

### ISSUE THREAD

**Issue by ihorml** [ant-design/ant-design#5790]
### What problem does this feature solve?
I'm creating a form which has `Input.Group` and there is no way to create Input.Group which support form validation using by Form.Item. I'm trying to do like:
```
<Input.Group>
  <Form.Item>getFieldDecorator(..)(<Input/>)</Form.Item>
  <Form.Item>getFieldDecorator(..)(<Input/>)</Form.Item>
</Input.Group>
```

### What does the proposed API look like?
I want make this work:
```
<Input.Group>
  <Form.Item>getFieldDecorator(..)(<Input/>)</Form.Item>
  <Form.Item>getFieldDecorator(..)(<Input/>)</Form.Item>
</Input.Group>
```

**Comment by benjycui** [comment]
I don't understand your use case, Form.Item is not required for validation.

**Comment by benjycui** [comment]
Close for no response.

**Comment by Xuhao** [comment]
> Form.Item is not required for validation.

我遇到了同样的问题，我需要一个`compact` 的 `Input.Group`,  里面通过getFieldDecorator放了2个 Input。2个 input 都有validation role，并且需要`hasFeedback`的支持。但是`hasFeedback`是放在`Form.Item` 上的，感觉是，为了能同时有2个`hasFeedback`，2个 input 就需要两个`Form.Item` 。

刚开始是这样做的:
```javascript
    <Form.Item label="评论内容" hasFeedback>
      <Input.Group compact>
        {getFieldDecorator('foreign_content', {
          rules: [{required: true, message: '请输入外文评论!'}]
        })(
          <Input
            style={{ width: '50%', height: '5rem' }}
            type="textarea"
            placeholder="在此输入外文评论"
            autosize={false} />
        )}
        {getFieldDecorator('chinese_content', {
          rules: [{required: true, message: '请输入中文评论!'}]
        })(
          <Input
            style={{ width: '50%', height: '5rem' }}
            type="textarea"
            placeholder="在此输入评论中文评论"
            autosize={false} />
        )}
      </Input.Group>
    </Form.Item>
```

结果发现 validation 只会显示第一个 input `foreign_content`的 validation 状态。并且 `hasFeedback`也只是显示第一个 input 的。

后来又尝试楼主的做法：
```javascript
    <Input.Group compact>
      <Form.Item hasFeedback>
        {getFieldDecorator('foreign_content', {
          rules: [{required: true, message: '请输入外文评论!'}]
        })(
          <Input
            style={{ width: '50%', height: '5rem' }}
            type="textarea"
            placeholder="在此输入外文评论"
            autosize={false} />
        )}
      </Form.Item>
      <Form.Item hasFeedback>
        {getFieldDecorator('chinese_content', {
          rules: [{required: true, message: '请输入中文评论!'}]
        })(
          <Input
            style={{ width: '50%', height: '5rem' }}
            type="textarea"
            placeholder="在此输入评论中文评论"
            autosize={false} />
        )}
      </Form.Item>
    </Input.Group>
```

发现2个`Form.Item`是无法`compact`的。

希望能看下`compact` 的 `Input.Group` 如何使用`getFieldDecorator`?

**Comment by benjycui** [comment]
自行 `getFieldErrors` 出来，放到对应的位置先吧。

**Comment by benjycui** [comment]
我问下视觉有没有好的建议。

**Comment by benjycui** [comment]
@ihorml According to me,  the whole Input.Group should be treat as one component. But you code means that you want to show error message for each item in Input.Group. I don't understand, could you show more information?

**Comment by ihorml** [comment]
@benjycui

This is my use case: 
<img width="424" alt="screenshot 2017-05-12 10 23 49" src="https://cloud.githubusercontent.com/assets/7770343/25987574/4600efe8-36fd-11e7-8ab0-8fd82f79c746.png">

**Comment by AmazingTurtle** [comment]
Having the same problem right now

**Comment by benjycui** [comment]
> Could you help me how can I implement this using ant.design features only without any js/css additions?

Nope.. In this use case, we must handle error status and messages manually.

**Comment by Sicria** [comment]
Also looking for this functionality. 

Is there any plans to make this possible? Should there maybe be a note or an example in the docs for future users as this does seem like a common use-case.

**Comment by mingeya** [comment]
Having the same problem right now.Is there any plans to make this possible? Thank you.

**Comment by slaby93** [comment]
Same problem now

**Comment by Knowledge91** [comment]
any update here?

**Comment by paradite** [comment]
For anyone looking to display a simple one line validation message, it is not very hard to do it within ant design framework. You can do it like this (TypeScript):

```tsx
class EditForm extends React.Component<EditFormProps, any> {
  state = {
    myValidateHelp: '',
    myValidateStatus: ''
  };

  private handleSubmit = (e: any) => {
    e.preventDefault();
    this.props.form.validateFields((err, values) => {
      if (!err) {
        this.setState({
          myValidateHelp: '',
          myValidateStatus: ''
        });
        console.log('Received values of form: ', values);
      } else {
        console.log(err);
        if (err.myA || err.myB) {
          this.setState({
            myValidateHelp: 'Please enter A and B',
            myValidateStatus: 'error'
          });
        }
      }
    });
  };

  public render() {
    const { getFieldDecorator } = this.props.form;
    const { myValidateHelp, myValidateStatus } = this.state;

    return (
      <div className="editForm">
          <Form.Item
            label="my ID"
            help={myValidateHelp}
            validateStatus={myValidateStatus}
          >
            <Group compact>
              {getFieldDecorator('myA', {
                rules: [
                  {
                    required: true,
                    message: 'Please input A!'
                  }
                ],
                initialValue: 'A'
              })(<Input />)}
              {getFieldDecorator('myB', {
                rules: [
                  {
                    required: true,
                    message: 'Please input B!'
                  }
                ],
                initialValue: 'B'
              })(<Input />)}
            </Group>
          </Form.Item>
        </Form>
      </div>
    );
  }
}

const WrappedEditForm = Form.create({})(EditForm);
export default WrappedEditForm;
```
If you need additional styling, you can use css and html in the `help` props.

**Comment by pavolgolias** [comment]
@benjycui @paradite 
I have the same problem with Input.Group and Form.Item.
I tried solution provided by Paradite, but there is problem because it is not good to use 2 `getFieldDecorator` in one `Form.Item` - I am getting this error message: Warning: `Form.Item` cannot generate `validateStatus` and `help` automatically, while there are more than one `getFieldDecorator` in it.

I also found some proposals is #11109 or #6045 (where I also found this [example](https://codesandbox.io/s/lppvny9n5l)), but there is also problem with multiple `getFieldDecorator` in one `Form.Item`. 

Is there any other solution how to properly use groups for various form items in form?

**Comment by paradite** [comment]
@pavolgolias I raised the problem with the error message as a separate issue: #14911
I am not sure if it will be fixed, but to me it is a minor issue since the users won't be looking at the console and the "error" is actually a "warning" that doesn't break anything: https://github.com/ant-design/ant-design/blob/b54f3674e4e93a6afb48348689067488b663a102/components/form/FormItem.tsx#L69

**Comment by paradite** [comment]
@pavolgolias it seems that the devs have listened to us and fixed the issue of warning message in #14911
🎉

**Comment by pavolgolias** [comment]
@paradite perfect, thanks for info ;)

**Comment by jjalonso** [comment]
Im suffering a similar issue. I have a group with a select and an input
I have 2 getFieldDecorator inside a form item and getting the error message, I also have the problem that is only showing the red error for the select but not for the input.

The input come in handleSubmit as an error, but not showing red message, only shown for the select when select is empty.

**Comment by kplachecki** [comment]
Still can't use multiple FieldDecorators in one Form.Item using Input.Group. Anyone solved it?

**Comment by Murillo2380** [comment]
@kplachecki I got this working by creating a "custom input" with [customized form controls](https://ant.design/components/form/#components-form-demo-customized-form-controls) and giving my custom input to `getFieldDecorator`.

Working example:

```js
/**
 * @param { {value: SalaryData, onChange: (SalaryData) => void} } props The component props
 */
const SalaryInput = ({ value: salaryData, onChange }) => {
    const triggerChange = changedData => {
        onChange(Object.assign({}, salaryData, changedData));
    };

    return (
        <InputGroup compact>
            <Select
                defaultValue={salaryData.isRanged ? "Between" : "Exact"}
                onChange={value =>
                    triggerChange({ isRanged: value === "Between" })
                }
            >
                <Option value="Between">Between</Option>
                <Option value="Exact">Exact</Option>
            </Select>
            <InputNumber
                min={1}
                style={{ width: 150, textAlign: "center" }}
                placeholder={salaryData.isRanged ? "Minimum" : "Amount"}
                onChange={salary => triggerChange({ salary })}
                defaultValue={salaryData.salary || undefined}
            />
            {salaryData.isRanged && (
                <Input
                    style={{
                        width: 30,
                        borderLeft: 0,
                        pointerEvents: "none",
                        backgroundColor: "#fff"
                    }}
                    placeholder="~"
                    disabled
                />
            )}
            {salaryData.isRanged && (
                <InputNumber
                    min={salaryData.salary + 1}
                    style={{ width: 150, textAlign: "center", borderLeft: 0 }}
                    placeholder="Maximum"
                    defaultValue={salaryData.maxSalary || undefined}
                    onChange={maxSalary => triggerChange({ maxSalary })}
                />
            )}

            <Select
                style={{ width: 100 }}
                showSearch
                placeholder="EUR"
                defaultValue={salaryData.currency}
                onChange={currency => triggerChange({ currency })}
            >
                {CURRENCY_CODES.map((currency, index) => (
                    <Option key={index} value={currency}>
                        {currency}
                    </Option>
                ))}
            </Select>
            <Input
                style={{
                    width: 100,
                    borderLeft: 0,
                    pointerEvents: "none",
                    backgroundColor: "#fff"
                }}
                placeholder="per year"
                disabled
            />
        </InputGroup>
    );
};

```

The **values** and **onChange** are provided by the Antd `getFieldDecorator`, no need for state managment (calling `props.onChange` already rerenders the component and updates any state from antd form).

```js
        <Form.Item
            label={title}
            style={{ fontWeight: "bold", marginTop }}
            key={key}
        >
            {getFieldDecorator(
                // initialValue: {...}, 
                rules: [{ required: true, message: "This field is required" }]
            })(
                (<SalaryInput /> )
            )}
        </Form.Item>
```

Result:
![image](https://user-images.githubusercontent.com/17615892/55495877-26c64580-563e-11e9-982f-805c1578098d.png)

Hope it helps!

**Comment by kplachecki** [comment]
@Murillo2380 thanks a lot!

**Comment by qiwang97** [comment]
@Murillo2380 thanks and it works. By the way , how to clear the values when i reset the form ?
i write a demo like this https://codesandbox.io/s/reverent-kalam-z7r58

**Comment by Murillo2380** [comment]
> @Murillo2380 thanks and it works. By the way , how to clear the values when i reset the form ?
> i write a demo like this https://codesandbox.io/s/reverent-kalam-z7r58

@qiwang97 Thanks for the demo! I did a `console.log` in the `value` prop from the custom input you created. The initial value you set when creating the form is being given when the form is reseted, but since it is a custom input, you need to forward the value to the inputs inside your custom input.

Check the working example here: https://codesandbox.io/s/resetable-custom-input-antd-ykhs2

For your case I changed this:

```jsx
<Form.Item>
          {getFieldDecorator("coordinate", {
            initialValue: ""
          })(<TestInput />)}
</Form.Item>
```
To this:

```jsx
<Form.Item>
          {getFieldDecorator("coordinate", {
            initialValue: ",,"
          })(<TestInput />)}
</Form.Item>
```

And inside your component I added:
```jsx
function CompactedInput({ value, onChange }, ref) {
    const [x, y, z] = value.split(",");
    // ...
    // ...
    <Input
          placeholder="coordinate-x"
          type="text"
          onChange={({ target: { value: val } }) => {
            triggerChange(val, 0);
          }}
          value={x /* add this */}
        />
```

There are some drawbacks, in this case, you can't use a comma as a valid character in your input, but since you are using the comma to separate every coordinate, I'm assuming it's ok to set the initial value as `",,"`. You can instead return an object as a value for your custom input like in my example above.

Hope it helps!

**Comment by qiwang97** [comment]
@Murillo2380  it works and thanks a lot for your help.

**Comment by rended** [comment]
Same problem, there are solutions in version 4.0.1?

**Comment by rended** [comment]
<Form.Item name={"zIndex"} label={translate('zIndex', ln)}>
    <Input.Group compact>
        <InputNumber min={0} max={8} defaultValue={0} size={size} />
        <Tooltip title={translate('bringsUp', ln)}>
            <Button type="primary" icon={<EyeOutlined />} size={size} onClick={bringsUp} />
        </Tooltip>
    </Input.Group>
</Form.Item>

**Comment by rended** [comment]
<Form.Item name={"zIndex"} label={translate('zIndex', ln)}>
    <Input.Group compact>
        <InputNumber min={0} max={8} defaultValue={0} size={size} />
        <Button type="primary" icon={<EyeOutlined />} size={size} />
    </Input.Group>
</Form.Item>

**Comment by rended** [comment]
<Form.Item name={"zIndex"} label={translate('zIndex', ln)}>
    <Input.Group compact>
        <InputNumber min={0} max={8} defaultValue={0} size={size} />
        <Button type="primary" icon={<EyeOutlined />} size={size} />
    </Input.Group>
</Form.Item>

**Comment by Murillo2380** [comment]
@rended I managed to get this working. Check [this comment](https://github.com/ant-design/ant-design/issues/5790?email_token=AEGMYFDA5JRKEVIWQQ4VWALRCMELRA5CNFSM4DH4NE42YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELN7LVQ#issuecomment-585247438) and [this other one](https://github.com/ant-design/ant-design/issues/5790?email_token=AEGMYFDA5JRKEVIWQQ4VWALRCMELRA5CNFSM4DH4NE42YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELN7LVQ#issuecomment-479563217) on this thread.

Basically you just need to create a component being a custom imput, give it to the antd and use the props `value` and `onChange` that are passed by the Antd form and inside your custom component you can use whatever you want.

I know the Form has changed with Antd 4. I didn't test it but you still should be able to do the same. Check the form rewrite section for how you can adapt my example to the new form api. #21656

**Comment by Ramthulasi** [comment]
How do i get the value of input?? i want to take this input value and set this value to state only on form submit finish.. not on the onValue change of the <Input> tag.. i dont want to use Form.item.. please help

<Modal
                centered
                title={<FormattedMessage id="payment.edit_amount" />}
                width={440}
                wrapClassName={ss.modalWrapper}
                visible={this.state.visible}
                onCancel={this.handleCancel}
                footer={null}
              >
                <Form 
                id="modified-amount"
                onFinish={this.onFinishAccountInfo}
                layout="vertical">
                  <div className={ss.paymentInfo}>
                    <div className={ss.editField}>
                      <Input
                        type="text"
                        defaultValue={this.state.amount}
                        autoFocus={true}
                        suffix="xyz"
                      />
                    </div>
                    <Card className={ss.alertCard}>
                      <Alert
                        message={<FormattedMessage id="global.attention" />}
                        banner
                        showIcon
                        icon={<ExclamationCircleFilled />}
                      />
                      <Paragraph>
                        <FormattedMessage id="payment.attention_desc" />
                      </Paragraph>
                    </Card>
                    <Button
                      form="modified-amount"
                      key="submit"
                      htmlType="submit"
                      type="primary"
                      className={ss.proceedPayBtn}
                      onClick={this.handleCancel}
                    >
                      <FormattedMessage id="user-account.saveChanges" />
                    </Button>
                  </div>
                </Form>
              </Modal>

**Comment by shubinla** [comment]
You can see:
https://ant.design/components/form-cn/#components-form-demo-complex-form-control
Add noStyle for Form.Item like:
`<Form.Item name="field" noStyle><Input /></Form.Item>`
